common/learning/notes.ts

```diff
@@ -0,0 +1,218 @@
+import { getLogger } from '../logger/logger';
+import { prisma } from '../prisma';
+import { User, getUser } from '../user';
+import { PrerequisiteError } from '../util/error';
+import { Prompts, prompt } from './llm/openai';
+import { LearningAssignment, LearningNote, LearningNoteType, LoKI, getAssignment } from './util';
+import { finishAssignment } from './assignment';
+
+const logger = getLogger(`LearningNotes`);
+
+// Constructs a note that is a reply to another note
+export const repliesTo = ({ assignmentId, topicId, id }: LearningNote) => ({ assignmentId, topicId, replyToId: id });
+
+export type LearningNoteCreate = Pick<LearningNote, 'text' | 'type'> & Partial<Pick<LearningNote, 'assignmentId' | 'topicId' | 'replyToId'>>;
+
+// Creates a new note on a topic or assignment, and then trigers an asynchronous reaction by the LLM
+export async function createNote(user: User | LoKI, data: LearningNoteCreate) {
+    if (!data.topicId && !data.assignmentId) {
+        throw new PrerequisiteError(`Note must be related to topic or assignment`);
+    }
+
+    const note = await prisma.learning_note.create({
+        data: {
+            ...data,
+            authorName: user === LoKI ? 'LoKI' : user.firstname,
+            authorID: user === LoKI ? null : user.userID,
+        },
+    });
+
+    logger.info(`User(${user === LoKI ? 'LoKI' : user.userID}) created LearningNote(${note.id})`);
+
+    // Fire and forget - async reaction by LLM
+    reactOnNote(note).catch((error) => logger.error(`Failed to react on Note`, error));
```

Comment:
Hm ... but how should we reply? Also an error might be temporary ... Dont have a good idea for now - also we cant throw an error as we reply asynchronously

Reply:
I'm not sure if this approach is the best fit. I was thinking of adding a `learning_note` with the author set to `System` or `Null`, which could be used to show some UI on the frontend. This would allow us to provide feedback to the user about the error, we could also include a retry button or something similar. 
However, it's not big deal right now.